### PR TITLE
Initial did-method-key implementation.

### PR DESCRIPTION
**Pull request by dmitrizagidulin**

### CODE REVIEW

**Comment by dlongley** [comment]
Some partial code to help with this PR:

https://gist.github.com/dlongley/4548cfa6d512df3c7d5d7740689cdb43

**Comment by dlongley** [comment]
We should also add `keyAgreement` to the dynamically generated keys and auto-produce an `X25519KeyAgreementKey` from the ed25519 key.

**Comment by dlongley** [comment]
This library can be used to convert a public ed25519 key to an x25519 key: https://github.com/dchest/ed2curve-js

**Comment by dlongley** [comment]
@dmitrizagidulin -- status update here? Do you think you could you get this working today?

File: lib/DidKeyDriver.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  async generate({keyType = 'Ed25519VerificationKey2018'}) {
+    const publicKey = await LDKeyPair.generate({type: keyType});
+    const did = 'did:key:' + publicKey.fingerprint();
+    return DidKeyDriver.didDoc({did, publicKey});
+  }
+
+  static didDoc({did, publicKey}) {
+    const {publicKeyBase58} = publicKey;
```
Comment (gannan08) [inline lib/DidKeyDriver.js]: This makes the assumption that we only support ED25519 keys that which uses the `publicKeyBase58` property and not RSA keys which use the `publicKeyPem`. Is it true that we are not supporting RSA keys?

Comment (dlongley) [reply lib/DidKeyDriver.js]: We are not supporting RSA keys for `did:key`.

File: lib/DidKeyDriver.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
+/*!
+ * Copyright (c) 2019 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+const {LDKeyPair} = require('crypto-ld');
+const base58 = require('bs58');
```
Comment (dmitrizagidulin) [inline lib/DidKeyDriver.js]: Note to self: figure out the difference between the `bs58` library and `forge`'s base58 export. (Both seem to be getting used interchangeably in our code?)

File: lib/convert.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+
+const base58 = require('bs58');
+const ed2curve = require('ed2curve');
+
+module.exports = {
+  convertToDhPublicKey(edPublicKeyBase58) {
```
Comment (dmitrizagidulin) [inline lib/convert.js]: TODO: Investigate if it makes sense to move this function to `crypto-ld` (my only hesitation is - not sure how heavy weight `ed2curve` lib is).

**Comment by dmitrizagidulin** [comment]
`keyAgreement` seems to be working, adding tests...

File: README.md [hunk 4]
```diff
@@ -1,2 +1,141 @@
-# did-key-driver
-A did-io driver for the DID "key" method
+# did:key method driver _(did-key-method)_
```
Comment (davidlehn) [inline README.md]: Repo name is `did-key-driver`, package.json says `did-io-key`, and here it says `did-key-method`.  They should match.  Related is maybe changing any of these drivers to use a form like "did-method-TYPE", similar to how other plugin systems work.

Comment (dmitrizagidulin) [reply README.md]: Oh right, thanks!

Comment (dmitrizagidulin) [reply README.md]: @davidlehn Changed the README and package.json to use `did-method-key` to match other drivers.
Can you or @dlongley rename this repo to `did-method-key` as well? (I don't seem to have admin)

Comment (dlongley) [reply README.md]: Renamed to `did-method-key`

File: README.md [hunk 5]
```diff
@@ -1,2 +1,141 @@
...
+    }
+  ]
+}
+```
+
+If you have an existing [`crypto-ld`](https://github.com/digitalbazaar/crypto-ld) 
+key pair, you can convert it to a `did:key` method DID Doc:
+
+```js
+const {toDidKeyMethodDoc} = require('did-key-driver');
```
Comment (davidlehn) [inline README.md]: Fix name.

File: lib/convert.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    const edPublicKeyBuffer = base58.decode(edPublicKeyBase58);
+    // Converts a 32-byte Ed25519 public key into a 32-byte Curve25519 key
+    // Returns null if the given public key in not a valid Ed25519 public key.
+    const dhPublicKeyBuffer = Buffer.from(ed2curve.convertPublicKey(edPublicKeyBuffer));
```
Comment (davidlehn) [inline lib/convert.js]: Buffer (and bs58 lib?) are Node only.  We need a browser friendly solution.

File: package.json [hunk 7]
```diff
@@ -0,0 +1,51 @@
+{
+  "name": "did-method-key",
+  "version": "0.0.1",
+  "description": "A did:key method driver for did-io and standalone use.",
+  "homepage": "http://github.com/digitalbazaar/did-method-key",
+  "author": {
+    "name": "Digital Bazaar, Inc.",
+    "email": "support@digitalbazaar.com",
+    "url": "http://digitalbazaar.com/"
```
Comment (davidlehn) [inline package.json]:
```suggestion
    "url": "https://digitalbazaar.com/"
```

File: package.json [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  },
+  "repository": {
+    "type": "git",
+    "url": "http://github.com/digitalbazaar/did-method-key"
```
Comment (davidlehn) [inline package.json]:
```suggestion
    "url": "https://github.com/digitalbazaar/did-method-key"
```

File: README.md (hunk 5 above)
Comment (dmitrizagidulin) [reply README.md]: Fixed, thanks.

File: lib/convert.js (hunk 6 above)
Comment (dmitrizagidulin) [reply lib/convert.js]: Fixed.

**Comment by dmitrizagidulin** [comment]
Changed to use browser-only libs.
Remaining steps:
- [ ] tests (wip)
- [ ] using a separate id (based off the X25519 fingerprint) for the keyAgreement section (wip)

**Comment by dmitrizagidulin** [comment]
@dlongley @davidlehn first draft ready to review.

File: README.md [hunk 9]
```diff
@@ -1,2 +1,144 @@
...
+  ],
+  "capabilityDelegation": [
+    "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH"
+  ],
+  "capabilityInvocation": [
+    "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH"
+  ],
+  "keyAgreement": [
+    {
+      "id": "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH#zBzoR5sqFgi6q3iFia8JPNfENCpi7RNSTKF7XNXX96SBY4",
```
Comment (dmitrizagidulin) [inline README.md]: @dlongley Just to double check - is this `keyAgreement.id` in the format that we expect? (Specifically, `did:key:<ed25519 fingerprint>#<curve25519 fingerprint>`)

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,11 @@
+language: node_js
+node_js:
+  - "8"
+  - "10"
```
Comment (davidlehn) [inline .travis.yml]: Add 12?

Comment (dmitrizagidulin) [reply .travis.yml]: @davidlehn node 12 is complicated, from what I understand. (has a bunch of branches, upstream - like this one https://github.com/digitalbazaar/crypto-ld/pull/34 )

File: README.md [hunk 11]
```diff
@@ -1,2 +1,154 @@
...
+Requires Node.js 8.3+
+
+To install locally (for development):
+
+```
+git clone https://github.com/digitalbazaar/did-method-key.git
+cd did-method-key
+npm install
+```
+
```
Comment (davidlehn) [inline README.md]: Could add docs for a direct npm install, assuming this will be published soon.

File: package.json [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  "dependencies": {
+    "crypto-ld": "^3.6.0",
+    "x25519-key-pair": "digitalbazaar/x25519-key-pair#initial"
+  },
+  "devDependencies": {
+    "chai": "^4.2.0",
+    "mocha": "^6.1.4"
+  },
+  "engines": {
+    "node": "^8.6.0"
```
Comment (davidlehn) [inline package.json]:
```suggestion
    "node": ">=8.6.0"
```

File: .travis.yml (hunk 10 above)
Comment (davidlehn) [reply .travis.yml]: Ok. I thought some of that was getting resolved. Could make a branch with just that change and poke travis periodically to see when it passes.

Comment (dmitrizagidulin) [reply .travis.yml]: Np, I'll give that a try.

File: README.md (hunk 11 above)
Comment (dmitrizagidulin) [reply README.md]: k!

File: .travis.yml (hunk 10 above)
Comment (davidlehn) [reply .travis.yml]: The crypto-ld node-12 branch is passing tests. Once published, at least that package won't be an issue.

File: README.md (hunk 9 above)
Comment (dlongley) [reply README.md]: Yes

**Comment by dlongley** [comment]
LGTM for release, @davidlehn? Need to keep any branches around until we've updated dependencies.

File: README.md (hunk 11 above)
Comment (dmitrizagidulin) [reply README.md]: Added

File: .travis.yml (hunk 10 above)
Comment (dmitrizagidulin) [reply .travis.yml]: Added support for Node 12 to travis.

**Pull request merged**